kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java

```diff
@@ -0,0 +1,455 @@
+      initPartitioner();
+      return partitioner.processStats(batchedOperatorStats);
+    } finally {
+      lastCheckTime = System.currentTimeMillis();
+    }
+  }
+
+  @Override
+  public Collection<Partition<AbstractKafkaInputOperator>> definePartitions(
+      Collection<Partition<AbstractKafkaInputOperator>> collection, PartitioningContext partitioningContext)
+  {
+    logger.debug("Define partitions");
+    initPartitioner();
+    return partitioner.definePartitions(collection, partitioningContext);
+  }
+
+  @Override
+  public void partitioned(Map<Integer, Partition<AbstractKafkaInputOperator>> map)
+  {
+    // update the last repartition time
+    lastRepartitionTime = System.currentTimeMillis();
+    initPartitioner();
+    partitioner.partitioned(map);
+  }
+
+
+  //---------------------------------------------setters and getters----------------------------------------
+  public void setInitialPartitionCount(int partitionCount)
+  {
+    this.initialPartitionCount = partitionCount;
+  }
+
+  public int getInitialPartitionCount()
+  {
+    return initialPartitionCount;
+  }
+
+  public void setClusters(String clusters)
+  {
+    this.clusters = clusters.split(";");
+  }
+
+  public String getClusters()
+  {
+    return Joiner.on(';').join(clusters);
+  }
+
+  public void setTopics(String... topics)
+  {
+    this.topics = topics;
+  }
+
+  public String[] getTopics()
+  {
+    return topics;
+  }
+
+  public void setConsumerWrapper(KafkaConsumerWrapper consumerWrapper)
+  {
+    this.consumerWrapper = consumerWrapper;
+  }
+
+  public KafkaConsumerWrapper getConsumerWrapper()
+  {
+    return consumerWrapper;
+  }
+
+  public void setAssignment(Set<AbstractKafkaPartitioner.PartitionMeta> assignment)
+  {
+    this.assignment = assignment;
+  }
+
+  public Set<AbstractKafkaPartitioner.PartitionMeta> getAssignment()
+  {
+    return assignment;
+  }
+
+  public Map<AbstractKafkaPartitioner.PartitionMeta, Long> getOffsetTrack()
+  {
+    return offsetTrack;
+  }
+
+  public void setStrategy(String policy)
+  {
+    this.strategy = PartitionStrategy.valueOf(policy.toUpperCase());
+  }
+
+  public PartitionStrategy getStrategy()
+  {
+    return strategy;
+  }
+
+  public void setInitialOffset(String initialOffset)
+  {
+    this.initialOffset = InitialOffset.valueOf(initialOffset.toUpperCase());
+  }
+
+  public InitialOffset getInitialOffset()
+  {
+    return initialOffset;
+  }
+
+  public String getApplicationName()
+  {
+    return applicationName;
+  }
+
+  public void setConsumerProps(Properties consumerProps)
+  {
+    this.consumerProps = consumerProps;
+  }
+
+  public Properties getConsumerProps()
+  {
+    return consumerProps;
+  }
+
+  public void setMaxTuplesPerWindow(int maxTuplesPerWindow)
+  {
+    this.maxTuplesPerWindow = maxTuplesPerWindow;
+  }
+
+  public int getMaxTuplesPerWindow()
+  {
+    return maxTuplesPerWindow;
+  }
+
+  public void setKeyDeserializer(Class<? extends Deserializer<K>> keyDeserializer)
```

Comment:
Shouldn't the serializers be protected instead? The derived class has to provide them, but not the user?